File: data/modules/lua/leaderboard_rpc.lua
--[[
 Copyright 2025 The Nakama Authors

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
--]]

local nk = require("nakama")

--[[
  Create all leaderboards persistent RPC function.
  This function creates global and per-game leaderboards with idempotent storage tracking.
--]]
local function create_all_leaderboards_persistent(context, payload)
  local token_url = "https://api.intelli-verse-x.ai/api/admin/oauth/token"
  local games_url = "https://gaming.intelli-verse-x.ai/api/games/games/all"

  local client_id = "54clc0uaqvr1944qvkas63o0rb"
  local client_secret = "1eb7ooua6ft832nh8dpmi37mos4juqq27svaqvmkt5grc3b7e377"

  local sort = "desc"
  local operator = "best"
  local reset_schedule = "0 0 * * 0" -- Weekly reset
  local collection = "leaderboards_registry"

  -- Fetch existing records to skip duplicates
  local existing_records = {}
  local success, records = pcall(nk.storage_read, {{
    collection = collection,
    key = "all_created",
    user_id = context.user_id or "system"
  }})

  if success and records and #records > 0 and records[1].value then
    existing_records = records[1].value
  else
    if not success then
      nk.logger_warn(string.format("Failed to read existing leaderboard records: %s", tostring(records)))
    end
  end

  -- Build set of existing IDs
  local existing_ids = {}
  for _, record in ipairs(existing_records) do
    existing_ids[record.leaderboardId] = true
  end

  local created = {}
  local skipped = {}

  -- Step 1: Request token
  nk.logger_info("Requesting IntelliVerse OAuth token...")
  local token_response
  success, token_response = pcall(nk.http_request, token_url, "post", {
    ["accept"] = "application/json",
    ["Content-Type"] = "application/json"
  }, nk.json_encode({
    client_id = client_id,
    client_secret = client_secret
  }))

  if not success then
    local error_msg = tostring(token_response)
    return nk.json_encode({
      success = false,
      error = string.format("Token request failed: %s", error_msg)
    })
  end

  if token_response.code ~= 200 then
    return nk.json_encode({
      success = false,
      error = string.format("Token request failed with code %d", token_response.code)
    })
  end

  local token_data
  success, token_data = pcall(nk.json_decode, token_response.body)
  if not success then
    return nk.json_encode({
      success = false,
      error = "Invalid token response JSON."
    })
  end

  local access_token = token_data.access_token
  if not access_token then
    return nk.json_encode({
      success = false,
      error = "No access_token in response."
    })
  end

  -- Step 2: Fetch game list
  nk.logger_info("Fetching onboarded game list...")
  local game_response
  success, game_response = pcall(nk.http_request, games_url, "get", {
    ["accept"] = "application/json",
    ["Authorization"] = string.format("Bearer %s", access_token)
  })

  if not success then
    local error_msg = tostring(game_response)
    return nk.json_encode({
      success = false,
      error = string.format("Game fetch failed: %s", error_msg)
    })
  end

  if game_response.code ~= 200 then
    return nk.json_encode({
      success = false,
      error = string.format("Game API responded with %d", game_response.code)
    })
  end

  local games
  success, games = pcall(function()
    local parsed = nk.json_decode(game_response.body)
    return parsed.data or {}
  end)

  if not success then
    return nk.json_encode({
      success = false,
      error = "Invalid games JSON format."
    })
  end

  -- Step 3: Create global leaderboard if missing
  local global_id = "leaderboard_global"
  if not existing_ids[global_id] then
    local create_success, _ = pcall(nk.leaderboard_create, global_id, true, sort, operator, reset_schedule, {
      scope = "global",
      desc = "Global Ecosystem Leaderboard"
    })
    if create_success then
      table.insert(created, global_id)
      table.insert(existing_records, {
        leaderboardId = global_id,
        scope = "global",
        createdAt = os.date("!%Y-%m-%dT%H:%M:%SZ")
      })
    else
      table.insert(skipped, global_id)
    end
  else
    table.insert(skipped, global_id)
  end

  -- Step 4: Create per-game leaderboards
  nk.logger_info(string.format("Processing %d games for leaderboard creation...", #games))
  for _, game in ipairs(games) do
    if game.id then
      local leaderboard_id = string.format("leaderboard_%s", game.id)
      if not existing_ids[leaderboard_id] then
        local create_success, _ = pcall(nk.leaderboard_create, leaderboard_id, true, sort, operator, reset_schedule, {
          desc = string.format("Leaderboard for %s", game.gameTitle or "Untitled Game"),
          gameId = game.id,
          scope = "game"
        })
        if create_success then
          table.insert(created, leaderboard_id)
          table.insert(existing_records, {
            leaderboardId = leaderboard_id,
            gameId = game.id,
            scope = "game",
            createdAt = os.date("!%Y-%m-%dT%H:%M:%SZ")
          })
        else
          table.insert(skipped, leaderboard_id)
        end
      else
        table.insert(skipped, leaderboard_id)
      end
    end
  end

  -- Step 5: Persist record of created leaderboards
  success, _ = pcall(nk.storage_write, {{
    collection = collection,
    key = "all_created",
    user_id = context.user_id or "system",
    value = existing_records,
    permission_read = 1,
    permission_write = 0
  }})

  if not success then
    nk.logger_error(string.format("Failed to write leaderboard records: %s", tostring(_)))
  end

  return nk.json_encode({
    success = true,
    created = created,
    skipped = skipped,
    totalProcessed = #games,
    storedRecords = #existing_records
  })
end

nk.register_rpc(create_all_leaderboards_persistent, "create_all_leaderboards_persistent")

File: data/modules/leaderboard_rpc.ts
// Copyright 2025 The Nakama Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

interface LeaderboardRecord {
    leaderboardId: string;
    gameId?: string;
    scope: string;
    createdAt: string;
}

let InitModule: nkruntime.InitModule = function (ctx: nkruntime.Context, logger: nkruntime.Logger, nk: nkruntime.Nakama, initializer: nkruntime.Initializer) {
    initializer.registerRpc("create_all_leaderboards_persistent", createAllLeaderboardsPersistent);
    logger.info("Leaderboard RPC registered successfully.");
};

function createAllLeaderboardsPersistent(ctx: nkruntime.Context, logger: nkruntime.Logger, nk: nkruntime.Nakama, payload: string): string {
    const tokenUrl = "https://api.intelli-verse-x.ai/api/admin/oauth/token";
    const gamesUrl = "https://gaming.intelli-verse-x.ai/api/games/games/all";

    const client_id = "54clc0uaqvr1944qvkas63o0rb";
    const client_secret = "1eb7ooua6ft832nh8dpmi37mos4juqq27svaqvmkt5grc3b7e377";

    const sort = "desc";
    const operator = "best";
    const resetSchedule = "0 0 * * 0"; // Weekly reset
    const collection = "leaderboards_registry";

    // Fetch existing records to skip duplicates
    let existingRecords: LeaderboardRecord[] = [];
    try {
        const records = nk.storageRead([{ collection, key: "all_created", userId: ctx.userId || "system" }]);
        if (records && records.length > 0 && records[0].value) {
            existingRecords = records[0].value as LeaderboardRecord[];
        }
    } catch (err) {
        logger.warn(`Failed to read existing leaderboard records: ${err}`);
    }

    const existingIds = new Set(existingRecords.map(r => r.leaderboardId));
    const created: string[] = [];
    const skipped: string[] = [];

    // Step 1: Request token
    logger.info("Requesting IntelliVerse OAuth token...");
    let tokenResponse;
    try {
        tokenResponse = nk.httpRequest(tokenUrl, "post", {
            "accept": "application/json",
            "Content-Type": "application/json"
        }, JSON.stringify({
            client_id,
            client_secret
        }));
    } catch (err) {
        const errorMsg = err instanceof Error ? err.message : String(err);
        return JSON.stringify({ success: false, error: `Token request failed: ${errorMsg}` });
    }

    if (tokenResponse.code !== 200) {
        return JSON.stringify({ success: false, error: `Token request failed with code ${tokenResponse.code}` });
    }

    let tokenData;
    try {
        tokenData = JSON.parse(tokenResponse.body);
    } catch (err) {
        return JSON.stringify({ success: false, error: "Invalid token response JSON." });
    }

    const accessToken = tokenData.access_token;
    if (!accessToken) {
        return JSON.stringify({ success: false, error: "No access_token in response." });
    }

    // Step 2: Fetch game list
    logger.info("Fetching onboarded game list...");
    let gameResponse;
    try {
        gameResponse = nk.httpRequest(gamesUrl, "get", {
            "accept": "application/json",
            "Authorization": `Bearer ${accessToken}`
        });
    } catch (err) {
        const errorMsg = err instanceof Error ? err.message : String(err);
        return JSON.stringify({ success: false, error: `Game fetch failed: ${errorMsg}` });
    }

    if (gameResponse.code !== 200) {
        return JSON.stringify({ success: false, error: `Game API responded with ${gameResponse.code}` });
    }

    let games;
    try {
        const parsed = JSON.parse(gameResponse.body);
        games = parsed.data || [];
    } catch (err) {
        return JSON.stringify({ success: false, error: "Invalid games JSON format." });
    }

    // Step 3: Create global leaderboard if missing
    const globalId = "leaderboard_global";
    if (!existingIds.has(globalId)) {
        try {
            nk.leaderboardCreate(globalId, true, sort, operator, resetSchedule, { scope: "global", desc: "Global Ecosystem Leaderboard" });
            created.push(globalId);
            existingRecords.push({ leaderboardId: globalId, scope: "global", createdAt: new Date().toISOString() });
        } catch (err) {
            skipped.push(globalId);
        }
    } else {
        skipped.push(globalId);
    }

    // Step 4: Create per-game leaderboards
    logger.info(`Processing ${games.length} games for leaderboard creation...`);
    for (const game of games) {
        if (!game.id) continue;
        const leaderboardId = `leaderboard_${game.id}`;
        if (existingIds.has(leaderboardId)) {
            skipped.push(leaderboardId);
            continue;
        }
        try {
            nk.leaderboardCreate(leaderboardId, true, sort, operator, resetSchedule, {
                desc: `Leaderboard for ${game.gameTitle || "Untitled Game"}`,
                gameId: game.id,
                scope: "game"
            });
            created.push(leaderboardId);
            existingRecords.push({
                leaderboardId,
                gameId: game.id,
                scope: "game",
                createdAt: new Date().toISOString()
            });
        } catch (err) {
            skipped.push(leaderboardId);
        }
    }

    // Step 5: Persist record of created leaderboards
    try {
        nk.storageWrite([{
            collection,
            key: "all_created",
            userId: ctx.userId || "system",
            value: existingRecords,
            permissionRead: 1,
            permissionWrite: 0
        }]);
    } catch (err) {
        const errorMsg = err instanceof Error ? err.message : String(err);
        logger.error(`Failed to write leaderboard records: ${errorMsg}`);
    }

    return JSON.stringify({
        success: true,
        created,
        skipped,
        totalProcessed: games.length,
        storedRecords: existingRecords.length
    });
}


